name: Merge main into PRs2

on:
  push:
    branches:
      - main

jobs:
  merge-main-into-prs:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout repository
      uses: actions/checkout@v4
      with:
        repository: ${{ github.event.pull_request.head.repo.full_name || github.repository }}
        ref: ${{ github.head_ref || github.ref }}
        fetch-depth: 0

    - uses: actions/setup-python@v5
      with:
        python-version: "3.11"
        cache: "pip" # caching pip dependencies

    - name: Install requirements
      shell: bash # for Windows compatibility
      run: |
        pip install pygithub

    - name: Merge main into PRs
      shell: python
      run: |
        from github import Github
        import os
        
        # Authenticate with the GitHub Token
        g = Github(os.getenv('GITHUB_TOKEN'))
        
        # Get the repository dynamically
        repo = g.get_repo(os.getenv('GITHUB_REPOSITORY'))
        
        # List all open pull requests
        open_pulls = repo.get_pulls(state='open', sort='created')
        
        for pr in open_pulls:
            try:                
                # Attempt to merge main into the PR branch
                message = f"Merge branch 'main' into {pr.head.ref} by [Ultralytics Actions](https://ultralytics.com)"
                merge_commit = pr.merge(pr.head.ref, 'main', commit_message=message)
                print(f'Merged main into {pr.head.ref} successfully.')
            except Exception as e:
                print(f'Could not merge main into {pr.head.ref}: {e}')
      env:
        GITHUB_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
        GITHUB_REPOSITORY: ${{ github.repository }}

